contrib/dlb/source/connection_balancer_impl.cc

```diff
@@ -0,0 +1,301 @@
+    }
+    if (dlb_enable_cq_epoll(rx_port, true, efd)) {
+      ExceptionUtil::throwEnvoyException(
+          fmt::format("dlb_enable_cq_epoll {}", errorDetails(errno)));
+    }
+    efds.push_back(efd);
+  }
+
+  if (dlb_launch_domain_alert_thread(domain, nullptr, nullptr)) {
+    ExceptionUtil::throwEnvoyException(
+        fmt::format("dlb_launch_domain_alert_thread {}", errorDetails(errno)));
+  }
+
+  if (dlb_start_sched_domain(domain)) {
+    ExceptionUtil::throwEnvoyException(
+        fmt::format("dlb_start_sched_domain {}", errorDetails(errno)));
+  }
+#endif
+
+  DlbConnectionBalanceFactorySingleton::initialize(this);
+
+  return std::make_shared<DlbConnectionBalancerImpl>();
+}
+
+DLBConnectionBalanceFactory::~DLBConnectionBalanceFactory() {
+  for (dlb_port_hdl_t port : rx_ports) {
+    if (dlb_disable_port(port)) {
+      ENVOY_LOG(error, "dlb_disable_port {}", errorDetails(errno));
+    }
+    if (dlb_detach_port(port) == -1) {
+      ENVOY_LOG(error, "dlb_detach_port {}", errorDetails(errno));
+    }
+  }
+  for (dlb_port_hdl_t port : tx_ports) {
+    if (dlb_disable_port(port)) {
+      ENVOY_LOG(error, "dlb_disable_port {}", errorDetails(errno));
+    }
+    if (dlb_detach_port(port) == -1) {
+      ENVOY_LOG(error, "dlb_detach_port {}", errorDetails(errno));
+    }
+  }
+  if (dlb_detach_sched_domain(domain) == -1) {
+    ENVOY_LOG(error, "dlb_detach_sched_domain {}", errorDetails(errno));
+  }
+
+  if (dlb_reset_sched_domain(dlb, domain_id) == -1) {
+    ENVOY_LOG(error, "dlb_reset_sched_domain {}", errorDetails(errno));
+  }
+
+  if (dlb_close(dlb) == -1) {
+    ENVOY_LOG(error, "dlb_close {}", errorDetails(errno));
+  }
+
+  for (int fd : efds) {
+    if (close(fd) == -1) {
+      ENVOY_LOG(error, "dlb close fd {}", errorDetails(errno));
+    }
+  }
+}
+
+void DlbBalancedConnectionHandlerImpl::setDlbEvent() {
+  auto listener = dynamic_cast<Envoy::Server::ActiveTcpListener*>(&handler_);
+
+  absl::MutexLock lock(&DlbConnectionBalanceFactorySingleton::get().lock);
+  dlb_event_ = listener->dispatcher().createFileEvent(
+      DlbConnectionBalanceFactorySingleton::get().efds[index_],
+      [this](uint32_t events) -> void { onDlbEvents(events); }, Event::FileTriggerType::Level,
+      Event::FileReadyType::Read);
+  dlb_event_->setEnabled(Event::FileReadyType::Read);
+}
+
+void DlbBalancedConnectionHandlerImpl::post(Network::ConnectionSocketPtr&& socket) {
+  absl::MutexLock lock(&DlbConnectionBalanceFactorySingleton::get().lock);
+
+  // The pointer will be casted to unique_ptr in onDlbEvents(), no need to consider free.
+  auto s = socket.release();
+
+  dlb_event_t events[1];
+  events[0].send.queue_id = DlbConnectionBalanceFactorySingleton::get().tx_queue_id;
+  events[0].send.sched_type = SCHED_UNORDERED;
+  events[0].adv_send.udata64 = reinterpret_cast<std::uintptr_t>(s);
+  int ret = dlb_send(DlbConnectionBalanceFactorySingleton::get().tx_ports[index_], 1, &events[0]);
```

Comment:
Would you mind to explain how do you do the balance? 

You create `DlbBalancedConnectionHandlerImpl` instance for each worker_id, `DlbConnectionBalancerImpl::pickTargetHandler` will return the same worker's `DlbBalancedConnectionHandlerImpl`, and at here seems you post to the same worker again. 

I'm not sure I understand correctly, but it seems always going to the same worker.

Reply:
DlbConnectionBalancerImpl::post will send socket to DLB hardware, and in DlbConnectionBalancerImpl::onDlbEvent, DLB hardware will trigger the corresponding worker to receive the socket, **DLB does all balance things**. In DlbConnectionBalancerImpl::setDlbEvent, it map rx_port and efd. Each worker has tx_port for send and rx_port for receive, all ports share one queue.

Since I need to do "send/recv" with DLB and cannot change original codes, I create DlbConnectionBalancerImpl here.

For more details, you can refer to https://networkbuilders.intel.com/solutionslibrary/queue-management-and-load-balancing-on-intel-architecture.

Reply:
thanks, better to explain how it works, since not everyone is coming from Intel :)

Reply:
Thanks for your nice tip :)